# Lambda School Data Science Module 124

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):` 

You should only truly need these three values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?


In [ ]:
# TODO - code!
#True Positive Rate
#true_positive_rate = 1.0
#Prior Probablity
#p_drunk = 1/1000
#False Positive Rate
#False_positive_rate = .08
#Complement Prior
#prob_drunk_prior = 1-p_drunk



#numerator = true_positive_rate * p_drunk
#denominator = true_positive_rate * p_drunk + false_positive_rate * prob_drunk_prior

#posterior_probability = numerator / denominator

#posterior_probability

In [ ]:
def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):
  prob_not_drunk = 1 - prob_drunk_prior

  numerator = true_positive_rate * prob_drunk_prior
  denominator = (true_positive_rate * prob_drunk_prior) + (false_positive_rate * prob_not_drunk)
  
  posterior_probability = numerator / denominator
  return posterior_probability


In [ ]:
results = prob_drunk_given_positive(1/1000, .08, 1)
print(results)

0.012357884330202669


In [ ]:
results = prob_drunk_given_positive(results, false_positive_rate, true_positive_rate)
print(results)

0.5028585632612479


In [ ]:
p_drunk = 1/1000

while p_drunk <= .95:
  p_drunk = prob_drunk_given_positive(p_drunk, .08, 1)
  print(p_drunk)

0.012357884330202669
0.13525210993291495
0.6615996951348605
0.9606895076105054


**1) What is the probability that a person is drunk after one positive breathalyzer test?**

1.2%

**2) What is the probability that a person is drunk after two positive breathalyzer tests?**

13.5%

**3) How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?**

4 tests are needed.

### 2) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).

Frequentists statistics looks at the probability of something without including any information other than the particular something we are trying to determine the probability of.  

Bayesian statistics also includes prior propabilities from previous observations in its calculations. 

### 3) Use the following Template to help come up with ideas for your Build Sprint Project: 



---

## Idea 1:  

### You
**What do you care about?**  Social Media

**What do you know about?**  Social Media

**What decisions do you face?** How, or if to use social media.




### Seven templates
Training Kit (https://learn.lambdaschool.com/ds/module/recedjanlbpqxic2r) explains the seven templates from Priceonomics.

Can you apply the templates to your topics?

**Geographic Variation:** Yes, I could look at how social media usage varies between different geographic locations

**Trend related to the news:** Social media and the news often seem to feed each other in their trends.

**Who does that?:**. Who uses what social media and how do they use it?

**Answering a question people care about:** Is social media usage bad for kids?

**Valuable to businesses:** How can using social media for marketing effect ROI?

**What's the most popular?:** what social media is used the most?

**Cost/Money rankings:** which sites make the most money, and why?

### Misconceptions

What misconceptions do people have about your topic?

That social media is just a tool for connecting with other people.

### Examples

What data storytelling example inspires you?

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

1) The first stat storytelling example that got me really into data science was a visual presentation (Youtube video)  on how divided congressional votes have become over the last 50 years.  https://www.youtube.com/watch?v=tEczkhfLwqM

2) Yes, comparing the data to other data sets to try and establish possible causality.

3) Yes, what is the correlation between the polarization observed in the voting record and polarization in our media?

4) Yes, you could compare a connection between any two groups over time using this visualization method.  

###Data

Where could you search for data about your topic?

Social Media studies are fairly common, and there are likely several available datasets already out there that I could play around with.  

---

## Idea 2:

### You
What do you care about? Music

What do you know about?  Music?

What decisions do you face?  What kinda music to play?



### Seven templates
Training Kit (https://learn.lambdaschool.com/ds/module/recedjanlbpqxic2r) explains the seven templates from Priceonomics.

Can you apply the templates to your topics?

**Geographic Variation:** What kind of music is most popular in different geographic locations?

**Trend related to the news:** How does news about artists drive record sales?

**Who does that?:** Who learns how to play music?

**Answering a question people care about:** Can music effect our mood?

**Valuable to businesses:** How can music effect marketing?

**What's the most popular?:** Look at streaming numbers 

**Cost/Money rankings:**. Which artists make the most money?

### Misconceptions

What misconceptions do people have about your topic?  Most misconceptions are mostly about the industry.

### Examples

What data storytelling example inspires you? 

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 & 3 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!